javalib/src/main/scala/java/math/BigDecimal.scala

```diff
@@ -0,0 +1,1761 @@
+
+      // The sign of the is fixed and the quotient will be saved in 'p2'
+      val p2 = if (q.signum() < 0) p.negate() else p
+
+      // Checking if the new scale is out of range
+      val newScale = safeLongToInt(diffScale + Math.max(k, l))
+      // k >= 0  and  l >= 0  implies that  k - l  is in the 32-bit range
+      val i = k - l
+      val p3 = if (i > 0) multiplyByFivePow(p2, i) else p2.shiftLeft(-i)
+      new BigDecimal(p3, newScale)
+    }
+  }
+
+  def divide(divisor: BigDecimal, mc: MathContext): BigDecimal = {
+    // Calculating how many zeros must be append to 'dividend'
+    // to obtain a  quotient with at least 'mc.precision()' digits
+
+    // In special cases it reduces the problem to call the dual method
+    if (mc.precision == 0 || this.isZero || divisor.isZero)
+      return this.divide(divisor)
+
+    val diffScale: Long = _scale.toLong - divisor._scale
+    val trailingZeros = mc.precision + 2L + divisor.approxPrecision() - approxPrecision()
+
+    val (quot: BigInteger, newScale0: Long) = {
+      if (trailingZeros > 0) {
+        // To append trailing zeros at end of dividend
+        val q  = getUnscaledValue.multiply(powerOf10(trailingZeros))
+        (q, diffScale + trailingZeros)
+      } else {
+        (getUnscaledValue, diffScale)
+      }
+    }
+
+    val qr: QuotAndRem[BigInteger] = quot.divideAndRemainder(divisor.getUnscaledValue)
+    val (integerQuot, newScale) = {
+      // Calculating the exact quotient with at least 'mc.precision()' digits
+      if (qr.rem.signum() != 0) {
+        // Checking if:   2 * remainder >= divisor ?
+        val compRem = qr.rem.shiftLeftOneBit().compareTo(divisor.getUnscaledValue)
+        val bi = BigInteger.valueOf(qr.quot.signum() * (5 + compRem))
+        (qr.quot.multiply(BigInteger.TEN).add(bi), newScale0 + 1)
+      } else {
+        // To strip trailing zeros until the preferred scale is reached
+        val lastPow = BigTenPows.length - 1
+
+        @inline
+        @tailrec
+        def loop(i: Int, iq: BigInteger, scale: Long): (BigInteger, Long) = {
+          if (!iq.testBit(0)) {
+            val qr: QuotAndRem[BigInteger] = iq.divideAndRemainder(BigTenPows(i))
+            if ((qr.rem.signum() == 0) && (scale - i >= diffScale))
+              loop(if (i < lastPow) i + 1 else i, qr.quot, scale - i)
+            else if (i != 1)
+              loop(1, iq, scale)
+            else
+              (iq, scale)
+          } else {
+            (iq, scale)
+          }
+        }
+
+        loop(1, qr.quot, newScale0)
+      }
+    }
+    // To perform rounding
+    new BigDecimal(integerQuot, safeLongToInt(newScale), mc)
+  }
+
+  def divideToIntegralValue(divisor: BigDecimal): BigDecimal = {
+    if (divisor.isZero)
+      throw new ArithmeticException("Division by zero")
+
+    val newScale: Long = this._scale.toLong - divisor._scale
+    val lastPow = BigTenPows.length - 1
+    val (integralValue: BigInteger, varScale) = {
```

Comment:
It means that this is translated to (roughly):
```scala
val tuple = { ... }
tuple match {
  case (integralValue: BigInteger, varScale) => ...
  case _ => throw new MatchError
}
```
so it won't check *statically* that the first item of the tuple is a `BigInteger`. This will typecheck whatever the first item's type is. Instead, it will test at *runtime* if it is a `BigInteger`, and if it's not, will throw a `MatchError`.